Docs/init gateway (#343)
it-at-m/refarch: docs/gateway.md

# API Gateway

RefArch gateway based on [Spring Cloud Gateway](https://spring.io/projects/spring-cloud-gateway).

## Usage

The gateway is released as container image `ghcr.io/it-at-m/refarch/refarch-gateway` and can be used either directly or
via the corresponding [Helm-Chart](https://github.com/it-at-m/helm-charts/tree/main/charts/refarch-gateway).

## Features

Beside the default functionality of Spring Cloud Gateway (i.e. routing) following features are preconfigured/provided:

- OAuth 2.0 based login
- Route protection
- CSRF protection with whitelist
- Cookie to JWT mapping for session management

> - Health and metrics endpoints

Reviewer: @simonhir What do you think about this?

> - Health and metrics endpoints

Author: Done

# API Gateway

RefArch gateway based on [Spring Cloud Gateway](https://spring.io/projects/spring-cloud-gateway).

::: info
In the previous version of the RefArch each project managed their own API gateway.

With the new version this should only occur very rarely and only if a very specific feature can't be implemented within the managed gateway.

> :::

Author: Done

## Usage

The gateway is released as container image `ghcr.io/it-at-m/refarch/refarch-gateway` and can be used either directly or
via the corresponding [Helm-Chart](https://github.com/it-at-m/helm-charts/tree/main/charts/refarch-gateway).

## Features

Beside the default functionality of Spring Cloud Gateway (i.e. routing) following features are preconfigured/provided:

- OAuth 2.0 based login
- Route protection
- CSRF protection with whitelist
- Cookie to JWT mapping for session management

> - Health and metrics endpoints

Author: Just took a short look and for instance all the filters are missing, but i don't really think that's a great place for just listing them.

> - Health and metrics endpoints

Author: Done

### Routing

Routes are configured via environment variables as listed under [Configuration](#configuration).

By default, routes require authentication through oAuth2 and manage the session between the client and gateway using
cookies.
The gateway then maps the session cookie to a JWT before routing it.

Beside the default behaviour there are some special route prefixes which are handled different:

- `/public/**`: All `OPTIONS` and `GET` requests are routed without security.
- `/clients/**`: Uses JWT for authenticating incoming requests instead of session cookies.

> ## Profiles

Reviewer: I´d move Profiles to a subheading inside of Configuration.

Author: Done

| `SPRING_CLOUD_GATEWAY_ROUTES_<index>_ID`                 | Id of a route definition.                                         | `backend`                                                               |
| `SPRING_CLOUD_GATEWAY_ROUTES_<index>_URI`                | The uri to route to if this route matches.                        | `http://backend-service:8080/`                                          |
| `SPRING_CLOUD_GATEWAY_ROUTES_<index>_PREDICATES_<index>` | Route predicates i.e. matcher.                                    | `Path=/api/backend-service/**`                                          |
| `SPRING_CLOUD_GATEWAY_ROUTES_<index>_FILTERS_<index>`    | List of filters applied to the route.                             | `RewritePath=/api/backend-service/(?<urlsegments>.*), /$\{urlsegments}` |
| `REFARCH_HAZELCAST_SERVICENAME`                          | Kubernetes service name for when using profile `hazelcast-k8s`.   |                                                                         |

> | `ALLOWED_ORIGINS_PUBLIC` (optional)                      | List of urls allowed as origin for public routes.                 | `https://*.example.com,http://localhost:*`                              |

Reviewer: ```suggestion
| `ALLOWED_ORIGINS_PUBLIC` (optional)                      | List of URIs allowed as origin for public routes.                 | `https://*.example.com,http://localhost:*`                              |
```

> | `ALLOWED_ORIGINS_CLIENTS` (optional)                     | List of urls allowed as origin for clients routes.                | `https://*.example.com,http://localhost:*`                              |

Reviewer: ```suggestion
| `ALLOWED_ORIGINS_CLIENTS` (optional)                     | List of URIs allowed as origin for clients routes.                | `https://*.example.com,http://localhost:*`                              |
```

> | `REFARCH_SECURITY_CSRFWHITELISTED_<index>` (optional)    | List of routes to disable csrf protection for.                    | `/example/**`                                                           |

Reviewer: ```suggestion
| `REFARCH_SECURITY_CSRFWHITELISTED_<index>` (optional)    | List of routes to disable CSRF protection for.                    | `/example/**`                                                           |
```

### Security

> For authentication via SSO, OAuth2 needs to be configured.

Reviewer: ```suggestion
For authentication via SSO, OAuth 2.0 needs to be configured.
```

- Route protection
- CSRF protection with whitelist
- Cookie to JWT mapping for session management

>   - Session synchronisation between multiple instances via Hazelcast (see sections [Profiles](#profiles) and [Hazelcast](#hazelcast))

Reviewer: Looks kind of strange, that this is a sub-bulletpoint because its the only occurence. I'd move it to its own bullet point.

Author: Don't think that looks strange and that's the correct place as exactly this is synchronized.
Would let it that way

Author: Done
